File: src/main/kotlin/com/linqingying/cangjie/psi/CjStringTemplateEntry.java
package com.linqingying.cangjie.psi;


import com.linqingying.cangjie.psi.stubs.CangJiePlaceHolderWithTextStub;
import com.intellij.lang.ASTNode;
import com.intellij.psi.stubs.IStubElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class CjStringTemplateEntry extends CjElementImplStub<CangJiePlaceHolderWithTextStub<? extends CjStringTemplateEntry>> {
    public static final CjStringTemplateEntry[] EMPTY_ARRAY = new CjStringTemplateEntry[0];

    public CjStringTemplateEntry(@NotNull ASTNode node) {
        super(node);
    }

    public CjStringTemplateEntry(
            @NotNull CangJiePlaceHolderWithTextStub<? extends CjStringTemplateEntry> stub,
            @NotNull IStubElementType elementType
    ) {
        super(stub, elementType);
    }

    @Nullable
    public CjExpression getExpression() {
        return findChildByClass(CjExpression.class);
    }
}

File: src/main/kotlin/com/huawei/cangjie/psi/CjStringTemplateEntry.kt
package com.huawei.cangjie.psi

import com.huawei.cangjie.psi.stubs.CangJiePlaceHolderWithTextStub
import com.intellij.lang.ASTNode
import com.intellij.psi.stubs.IStubElementType


abstract class CjStringTemplateEntry : CjElementImplStub<CangJiePlaceHolderWithTextStub<out CjStringTemplateEntry > > {
    constructor(node: ASTNode) : super(node)

    constructor(
        stub: CangJiePlaceHolderWithTextStub<out CjStringTemplateEntry >,
        elementType: IStubElementType<*, *>
    ) : super(stub, elementType)

    val expression: CjExpression?
        get() = findChildByClass(CjExpression::class.java)

    companion object {
        val EMPTY_ARRAY: Array<CjStringTemplateEntry?> = arrayOfNulls(0)
    }
}
